drivers/hmis/app/graphql/concerns/config_tool_permission_helper.rb

```diff
@@ -0,0 +1,16 @@
+#  Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+#  License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+#
+
+module ConfigToolPermissionHelper
+  extend ActiveSupport::Concern
+
+  def ensure_form_role_permission(role)
```

Comment:
I'm happy to refactor to put this role-check permission on the `User` model (or, the [UserPermissions](https://github.com/greenriver/hmis-warehouse/blob/4c4a9e76d153de126451670502e2db9b30a60b10/app/models/concerns/user_permissions.rb#L4) concern?), but want to note that I specifically did not make `ensure_form_role_permission` check for `can_manage_forms`, because `ensure_form_role_permission` is also used to check access (not edit) permissions for form definitions/identifiers, which the user only needs `can_configure_data_collection` to view. (This is a bit related to my other comment [here](https://github.com/greenriver/hmis-warehouse/pull/4531#discussion_r1676377813))

Reply:
Maybe the naming is more confusing than it needs to be? I'll think more on the permission names and at the very least add some clarifying comments

Reply:
Maybe the `UserPermissions` concern should get 2 new methods, `can_view_form_role` and `can_edit_form_role`

Reply:
I guess that naming still doesn't fully clarify it, because `can_configure_data_collection` is an editing permission. So e.g. in the `CreateFormRule` mutation we want to check `can_configure_data_collection` combined with `can_administrate_config` if the `role` is one of the restricted roles, and we don't want to just call that `can_view_form_role`, because they are making a mutation, but it doesn't require full `can_manage_forms` permission either.

Reply:
I also think there is an advantage to checking `can_manage_forms?` separately before `can_administrate_config` in the cases where we have to query for a definition/instance in order to get the form role. If we bundle them together and only check after querying for the definition, we may return 'not found' when the user should actually see 'access denied', which would be a permission leak, right?

Reply:
Alright gotcha. It does make it confusing. I think I'm a little confused by the way that `ensure_form_role_permission` is used separately from the permission checks. That makes the mutation code hard to follow, for me anyway.

Maybe I am over-simplifying here, but can we just create a separate method for the "configure data collection" perm?
```
user.can_manage_forms_for_role?(role)?  // true if can_manage_forms && role is permitted
user.can_configure_data_collection_for_role?(role)?  // true if can_configure_data_collection && role is permitted
```

Then we only need to wrap our heads around those 2 permissions. This is similar conceptually to how permissions apply differently based on the project, so we have a mental model for it. (Eg the user has this permission in the context of this "entity", which in this case is a role, sort of).


```
# form mutations
access_denied! unless current_user.can_manage_forms_for_role?(input.role)

# form rule mutations
access_denied! unless current_user.can_configure_data_collection_for_role(definition.role)

```

Reply:
Sorry I didn't see your last 3 comments when I posted @martha . Reading..

Reply:
So, I _think_ my suggestion addresses your concern mentioned https://github.com/greenriver/hmis-warehouse/pull/4531#discussion_r1676396640, but let me know if it does not.

And to address this one:

> I also think there is an advantage to checking can_manage_forms? separately before can_administrate_config in the cases where we have to query for a definition/instance in order to get the form role. If we bundle them together and only check after querying for the definition, we may return 'not found' when the user should actually see 'access denied', which would be a permission leak, right?

That's a good thought about the leak, it's true it exposes some more information. In this case, that information is just about the existence of a certain form identifier, or form role, or form id. None of those are really leaks that concern me. They will see the these forms when interacting with the application, as do all users. Forms are essentially _config_, not data. Anyone can see it, but only certain people can edit it. That is the philosophy I would stick with, anyway. (As you can see, we have _no_ permission checks on [this](https://github.com/greenriver/hmis-warehouse/blob/51b622d79a3b74bcc64c304fa3539ee653554cec/drivers/hmis/app/graphql/types/hmis_schema/query_type.rb#L178) or [this](https://github.com/greenriver/hmis-warehouse/blob/51b622d79a3b74bcc64c304fa3539ee653554cec/drivers/hmis/app/graphql/types/hmis_schema/query_type.rb#L232) query for fetching FDs. At some point we decided to add permission checks for the FD queries that are used just for configuration, mainly just as an implementation note that they _should only_ be used for configuration screens, not for application screens. That may have been a confusing choice.

Reply:
Okay, thanks for this and sorry for getting so bogged down, thanks for helping me sort it out! I think we settled on similar ideas at the same time where `UserPermissions` gets 2 more permissions that each take `role` as an input, one for management and one for data configuration. I think it's much more clear now

Reply:
Ok, great, thanks!